docs: add an Ubuntu 26.04 migration guide (#2673)
canonical/charmcraft: docs/howto/migrate-bases/change-to-ubuntu@26.04.rst

.. _howto-change-to-ubuntu-26-04:

.. meta::

>     :description: Learn how to migrate a charm to the Ubuntu 26.04 base in Charmcraft, including updating from the bases key to base and platforms keys.

Reviewer: ```suggestion
    :description: How to migrate a charm to the Ubuntu 26.04 LTS base in Charmcraft, including using the new platform definitions.
```

I finally have updated guidance about these: https://library.canonical.com/documentation/documentation-practice/html-metadata/set-useful-html-meta-descriptions#write-a-meta-description-1

Author: Should we retire https://canonical-starflow.readthedocs-hosted.com/how-to/add-a-page-meta-description/ then?

Reviewer: Working on it!

Change to Ubuntu 26.04

> ======================

Reviewer: ```suggestion
Change to the Ubuntu 26.04 LTS base
===================================
```

Our branding standards require this.

> This guide describes the process for migrating a charm from an older base to Ubuntu 26.04.

Reviewer: ```suggestion
This guide describes the process for migrating a charm from an older base to Ubuntu 26.04 LTS.
```

Removed features

> ----------------

Reviewer: ```suggestion
Migrate from the charm plugin
-----------------------------
```

The charm plugin isn't available for the ``ubuntu@26.04`` base. Before changing the

> base, follow one of the plugin migration guides:

Reviewer: ```suggestion
The charm plugin isn't available for the ``ubuntu@26.04`` base. If your charm uses it,
before changing the base, switch to one of the replacement plugins:
```

> base, follow one of the plugin migration guides:

Reviewer: Is there some way we can help the user decide which one? Is there anything to suggest beyond preference?

Author: I mean... *my* preference is for everyone to use the uv plugin, but if they're already using poetry or something else to manage their dependencies we shouldn't be forcing them to switch.

Reviewer: I suppose it would be based on their existing packaging.

- :ref:`howto-migrate-to-uv`
- :ref:`howto-migrate-to-python`
- :ref:`howto-migrate-to-poetry`

``bases`` key

> -------------

Reviewer: ```suggestion
Update the base
---------------
```

Charms currently using Ubuntu 22.04 or older might use the ``bases`` key in

> ``charmcraft.yaml``. Charmcraft doesn't support this with Ubuntu 26.04, instead relying

Reviewer: ```suggestion
Charms currently using Ubuntu 22.04 LTS or older might use the ``bases`` key in
``charmcraft.yaml``. Charmcraft doesn't support this with Ubuntu 26.04 LTS, instead relying
```

on the ``base`` and ``platforms`` keys. The ``base`` key controls which Ubuntu release
the charm uses, while the ``platforms`` key decides the CPU architectures of both the
build machine and the production machine.

For a charm with a ``bases`` key as follows:

.. code-block:: yaml

> 

Reviewer: ```suggestion
    :caption: charmcraft.yaml

```

    bases:
      - build-on:
          - name: ubuntu
            channel: "22.04"
        run-on:
          - name: ubuntu
            channel: "22.04"

Replace the ``bases`` key with:

> 

Reviewer: ```suggestion
    :caption: charmcraft.yaml

```

.. code-block:: yaml

    base: ubuntu@26.04
    platforms:
      amd64:
      arm64:
      riscv64:
      s390x:

> The :ref:`reference-platforms` has all the details about the ``platforms`` syntax,

Reviewer: ```suggestion
:ref:`reference-platforms` has all the details about the ``platforms`` syntax,
```

Change to the Ubuntu 26.04 LTS base
===================================

This guide describes the process for migrating a charm from an older base to Ubuntu 26.04 LTS.

Migrate from the charm plugin
-----------------------------

The charm plugin isn't available for the ``ubuntu@26.04`` base. If your charm uses it,
before changing the base, switch to one of the replacement plugins:

- :ref:`howto-migrate-to-uv`
- :ref:`howto-migrate-to-python`
- :ref:`howto-migrate-to-poetry`

Update the base
---------------

Charms currently using Ubuntu 22.04 LTS or older might use the ``bases`` key in
``charmcraft.yaml``. Charmcraft doesn't support this with Ubuntu 26.04 LTS, instead relying
on the ``base`` and ``platforms`` keys. The ``base`` key controls which Ubuntu release
the charm uses, while the ``platforms`` key decides the CPU architectures of both the
build machine and the production machine.

For a charm with a ``bases`` key as follows:

.. code-block:: yaml
    :caption: charmcraft.yaml

    bases:
      - build-on:
          - name: ubuntu
            channel: "22.04"
        run-on:
          - name: ubuntu
            channel: "22.04"

Replace the ``bases`` key with:

> .. code-block:: yaml

Author: ```suggestion
.. code-block:: yaml
    :caption: charmcraft.yaml
```

.. _howto-change-to-ubuntu-26-04:

.. meta::
    :description: How to migrate a charm to the Ubuntu 26.04 LTS base in Charmcraft, including using the new platform definitions.

Change to the Ubuntu 26.04 LTS base
===================================

> This guide describes the process for migrating a charm from an older base to Ubuntu 26.04 LTS.

Reviewer 2: ```suggestion
This guide describes the process for migrating a charm from a lower base to Ubuntu 26.04 LTS.
```

Migrate from the charm plugin
-----------------------------

> The charm plugin isn't available for the ``ubuntu@26.04`` base. If your charm uses it,

Reviewer 2: ```suggestion
The Charm plugin isn't available for the Ubuntu 26.04 LTS base. If your charm uses it,
```
> The charm plugin

Capitalization was inconsistent with the rest of the docs.

> the `ubuntu@26.04` base

This is the only time this appears in this page's prose, so it feels slightly inconsistent.

Author: I'll do whatever you and @medubelko agree upon here.

Reviewer: @jahn-junior I originally urged ubuntu@26.04 like in the other crafts, but Alex is concerned that charm users specifically might not have the same awareness of bases as other developers, and will think of their need as "working with Ubuntu 26.04".

Reviewer 2: @medubelko

> Alex is concerned that charm users specifically might not have the same awareness of bases as other developers, and will think of their need as "working with Ubuntu 26.04"

This seems like it supports my suggestion? I'm not sure I understand.

Since this is the only instance of "the `<base>` base", I think either is fine. This isn't a detail I'd hold this up over. :+1:

Reviewer: Right, let's go with `Ubuntu 26.04 LTS`.

> before changing the base, switch to one of the replacement plugins:

Reviewer 2: ```suggestion
The charm plugin isn't available for the ``ubuntu@26.04`` base. If your charm uses it,
switch to one of the replacement plugins:
```
The order of the sections makes this clear.

- :ref:`howto-migrate-to-uv`
- :ref:`howto-migrate-to-python`
- :ref:`howto-migrate-to-poetry`

Update the base
---------------

Charms currently using Ubuntu 22.04 LTS or older might use the ``bases`` key in
``charmcraft.yaml``. Charmcraft doesn't support this with Ubuntu 26.04 LTS, instead relying
on the ``base`` and ``platforms`` keys. The ``base`` key controls which Ubuntu release
the charm uses, while the ``platforms`` key decides the CPU architectures of both the

> build machine and the production machine.

Reviewer 2: ```suggestion
Charms built with Ubuntu 22.04 LTS or lower might use the ``bases`` key in their project
file. This key is not supported by the Ubuntu 26.04 LTS base and must be replaced with
the ``base`` and ``platforms`` keys. The ``base`` key declares which Ubuntu release the
charm uses, while the ``platforms`` key declares the CPU architectures of the build and
production machines.
```

For a charm with a ``bases`` key as follows:

.. code-block:: yaml
    :caption: charmcraft.yaml

    bases:
      - build-on:
          - name: ubuntu
            channel: "22.04"
        run-on:
          - name: ubuntu
            channel: "22.04"

Replace the ``bases`` key with:

.. code-block:: yaml
    :caption: charmcraft.yaml

    base: ubuntu@26.04
    platforms:
      amd64:
      arm64:
      riscv64:
      s390x:

:ref:`reference-platforms` has all the details about the ``platforms`` syntax,

> including the grammar for specifying multiple bases and architectures.

Reviewer: Based on a discussion from today and anticipation of https://github.com/canonical/craft-application/pull/1069:

```suggestion
including the grammar for specifying multiple bases and architectures.

Update part names
-----------------

If you update a charm to use the Ubuntu 26.04 LTS base, then you must also verify its
part names. Part names on 26.04 and later bases can't contain any forward slashes (/).
We recommend replacing them with a hyphen (-):

.. code-block:: diff
    :caption: charmcraft.yaml

     base: ubuntu@26.04

     # ...

     parts:
    -  my/part:
    +  my-part:
```

> including the grammar for specifying multiple bases and architectures.

Reviewer 2: ```suggestion
:ref:`reference-platforms` has all the details about the ``platforms`` key,
including the syntax for specifying multiple bases and architectures.
```
To avoid having "syntax" and "grammar" in the same sentence.
